subprojects/process-services/src/main/java/org/gradle/process/internal/health/memory/OsMemoryStatus.java

```diff
@@ -22,11 +22,29 @@
 public interface OsMemoryStatus {
     /**
      * @return OS total memory in bytes
+     * @deprecated Use {@link #getPhysicalMemory()}{@code .}{@link OsMemoryCategory.Limited#getTotal() getTotal()} instead.
      */
+    @Deprecated
     long getTotalPhysicalMemory();
 
     /**
      * @return OS free memory in bytes
+     * @deprecated Use {@link #getPhysicalMemory()}{@code .}{@link OsMemoryCategory.Limited#getFree() getFree()} instead.
      */
+    @Deprecated
     long getFreePhysicalMemory();
+
+    /**
+     * Get the physical memory information.
+     *
+     * @return the physical memory information
+     */
+    OsMemoryCategory.Limited getPhysicalMemory();
```

Comment:
🤔 What if, instead of creating this distinction and requiring the consumer to check for `Limited` or not, we move the "do we need to free memory" question into OsMemoryStatus?  Something like `RequiredMemory getRequiredFreeMemoryFor(long memoryAmountInBytes, long minFreePercentage)`.  `RequiredMemory` would provide both the amount of required free memory bytes to satisfy the request and the name of the memory region that requires it.  Or it would return RequiredMemory.NONE (0 required bytes) if no free memory was required.  We'd probably also need a `boolean isAboveFreeMemoryThreshold(minFreePercentage)` or some such for the daemon expiration strategy, too.

What I like about this is that it moves all of the complexity of calculating memory thresholds into one place that can easily be tested (and maybe make some of it more reusable) rather than splitting it across different consumers.  It also lowers the overhead of `OsMemoryStatus` so that the consumer doesn't have to know about physical vs virtual memory - he can just treat everything the same and ask a simple question (i.e. I need X memory, do I need to free some up or not?)

Reply:
The only question I would have is if anything outside of GBT uses the exposed memory information, since I think it would be odd to only have the physical memory exposed in that case and would rather remove those functions.